File: iOS/Code/Xamarin/bugTrap/bugTrapKit/OverlayWindow/BtOverlayViewController.cs
using System;
using UIKit;
using CoreGraphics;

namespace bugTrapKit
{
	public class BtOverlayViewController : BtViewController
	{
		public bool InSdk { get; set; }

		public UITapGestureRecognizer TapGesture { get; set; }

		public UIPanGestureRecognizer PanGesture { get; set; }

		public UIImageView Trigger { get; set; }

		CGPoint triggerCenter;

		CGRect triggerCenterBoundsCache;

		public CGRect TriggerCenterBounds {
			get {
				if (triggerCenterBoundsCache == CGRect.Empty) {
					triggerCenterBoundsCache = new CGRect (UIScreen.MainScreen.Bounds.GetMidX() - 20f, UIScreen.MainScreen.Bounds.GetMidY() - (InSdk ? 20f : 100f), 40f, 40f);
				}
				return triggerCenterBoundsCache;
			}
		}

		static CGRect triggerBounds = UIScreen.MainScreen.Bounds.Inset(20f, 20f);

		public BtOverlayViewController ()
		{
			View = new BtOverlayView { Frame = UIScreen.MainScreen.Bounds };

			InitTrigger();

			InitGestureRecognizers();
		}

		//		public void EnsureBuildUp ()
		//		{
		//			base.BuildUp ();
		//		}

		void InitTrigger ()
		{
			Trigger = new UIImageView {
				Frame = TriggerCenterBounds,
				ContentMode = UIViewContentMode.ScaleAspectFit,
				Image = UIImage.FromBundle("i_trigger"),
				Alpha = 0.0f
			};

			// Trigger.Layer.CornerRadius = 20f;
			// Trigger.Layer.MasksToBounds = true;

			Trigger.UserInteractionEnabled = true;

			View.AddSubview(Trigger);
		}

		void InitGestureRecognizers ()
		{
			TapGesture = new UITapGestureRecognizer (HandleTapGestureAsync);
			PanGesture = new UIPanGestureRecognizer (HandlePanGesture);
		}

		public override void ViewWillAppear (bool animated)
		{
			base.ViewWillAppear(animated);

			Trigger.AddGestureRecognizer(TapGesture);
			Trigger.AddGestureRecognizer(PanGesture);
		}

		public async override void ViewDidAppear (bool animated)
		{
			base.ViewDidAppear(animated);

			var frame = new CGRect (4f, View.Bounds.Bottom - 94f, 40f, 40f);

			await UIView.AnimateAsync(1.0f, () => {
				Trigger.Alpha = 1.0f;
			});

			await UIView.AnimateAsync(0.8f, () => {
				Trigger.Frame = frame;
			});
		}

		public override void ViewWillDisappear (bool animated)
		{
			Trigger.RemoveGestureRecognizer(TapGesture);
			Trigger.RemoveGestureRecognizer(PanGesture);

			base.ViewWillDisappear(animated);
		}

		void HandlePanGesture (UIPanGestureRecognizer pan)
		{
			switch (pan.State) {
			case UIGestureRecognizerState.Changed:
				CGPoint translation = pan.TranslationInView(View);
				var center = new CGPoint (triggerCenter.X + translation.X, triggerCenter.Y + translation.Y);
				if (triggerBounds.Contains(center)) {
					Trigger.Center = center;
				}

				return;
			case UIGestureRecognizerState.Began:
				triggerCenter = Trigger.Center;
				return;
			case UIGestureRecognizerState.Ended:
				triggerCenter = CGPoint.Empty;
				return;
			}
		}

		async void HandleTapGestureAsync (UIGestureRecognizer tap)
		{
			if (tap.State == UIGestureRecognizerState.Recognized) {

				UIImage image;

				UIGraphics.BeginImageContextWithOptions(UIScreen.MainScreen.Bounds.Size, true, 0);

				UIApplication.SharedApplication.Windows[0].DrawViewHierarchy(UIScreen.MainScreen.Bounds, false);

				image = UIGraphics.GetImageFromCurrentImageContext();

				UIGraphics.EndImageContext();

//				image.SaveToPhotosAlbum((i, e) => {
//					
//				});

				await TrapState.Shared.AddSnapshotImageForSdk(image);

				// BugTrapState.SnapshotImage = image;

				var storyboard = UIStoryboard.FromName("bugTrapKit", NibBundle);

				var navController = storyboard.Instantiate<BtAnnotateImageNavigationController>();

				PresentViewController(navController, false, null);
			}
		}

		public void ToggleTriggerVisibility ()
		{
			Trigger.Hidden = !Trigger.Hidden;
		}
	}
}

File: iOS/Code/Swift/bugTrap/bugTrapKit/Window/BtOverlayViewController.swift
//
//  BtOverlayViewController.swift
//  bugTrap
//
//  Created by Colby L Williams on 8/23/14.
//  Copyright (c) 2014 bugTrap. All rights reserved.
//

import Foundation
import UIKit

class BtOverlayViewController: UIViewController {
	
	var tapGesture = UITapGestureRecognizer()
	var panGesture = UIPanGestureRecognizer()
	
	var trigger = UIImageView()
	
	var triggerCenter: CGPoint = CGPoint.zeroPoint
//	var triggerCenterBoundsCache = CGRect(x: UIScreen.mainScreen().bounds.midX - 20, y: UIScreen.mainScreen().bounds.midY - 20, width: 40, height: 40)
	
	let triggerCenterBounds = CGRect(x: UIScreen.mainScreen().bounds.midX - 20, y: UIScreen.mainScreen().bounds.midY - 20, width: 40, height: 40)
	
	var triggerBounds = CGRectInset(UIScreen.mainScreen().bounds, 20, 20)
	
	
	init() {
		super.init(nibName: nil, bundle: nil)
		
		view = BtOverlayView()
		
		initTrigger()
		
		initGestureRecognizers()
	}
	
	required init?(coder aDecoder: NSCoder) {
		super.init(coder: aDecoder)
	}
	
	
	override func viewWillAppear(animated: Bool) {
		super.viewWillAppear(animated)
		
		trigger.addGestureRecognizer(tapGesture)
		trigger.addGestureRecognizer(panGesture)
	}
	
	
	override func viewDidAppear(animated: Bool) {
		super.viewDidAppear(animated)
		
		let frame = CGRect(x: 4, y: view.bounds.height - 94, width: 40, height: 40)
		
		UIView.animateWithDuration(1, animations: {
			self.trigger.alpha = 1
		}) { _ in
			UIView.animateWithDuration(0.8) {
				self.trigger.frame = frame
			}
		}
	}
	
	
	override func viewWillDisappear(animated: Bool) {

		trigger.removeGestureRecognizer(tapGesture)
		trigger.removeGestureRecognizer(panGesture)
		
		super.viewWillDisappear(animated)
	}
	
	
	func initTrigger () {
		
		trigger.frame = triggerCenterBounds
		trigger.contentMode = .ScaleAspectFit
		trigger.image = UIImage(named: "i_trigger")
		trigger.alpha = 0;
		
		trigger.userInteractionEnabled = true
		
		view.addSubview(trigger)
	}
	
	
	func initGestureRecognizers() {
		
		tapGesture.addTarget(self, action: "handleTapGesture:")
		panGesture.addTarget(self, action: "handlePanGesture:")
	}

	
	func handleTapGesture (tap: UITapGestureRecognizer) {
		
		if tap.state == .Recognized {
			
			var image: UIImage
			
			UIGraphicsBeginImageContextWithOptions(UIScreen.mainScreen().bounds.size, true, 0)
			
			UIApplication.sharedApplication().windows[0].drawViewHierarchyInRect(UIScreen.mainScreen().bounds, afterScreenUpdates: false)
			
			image = UIGraphicsGetImageFromCurrentImageContext()
			
			UIGraphicsEndImageContext()
			
			TrapState.Shared.addSnapshotImageForSdk(image) { _ in
				
				let kitStoryboard = UIStoryboard(name: "bugTrapKit", bundle: self.nibBundle)
				
				let navController = kitStoryboard.instantiateViewControllerWithIdentifier("BtAnnotateImageNavigationController")

				self.presentViewController(navController, animated: false, completion: nil)
			}
		}
	}
	
	
	func handlePanGesture (pan: UIPanGestureRecognizer) {
		
		switch pan.state {
		case .Changed:

			let translation = pan.translationInView(view)
			
			let center = CGPoint(x: triggerCenter.x + translation.x, y: triggerCenter.y + translation.y)
			
			if triggerBounds.contains(center) {
				trigger.center = center
			}
			
		case .Began: triggerCenter = trigger.center
			
		case .Ended: triggerCenter = CGPoint.zeroPoint
			
		default: return
		}
	}
	
	func toggleTriggerVisibility() {
		trigger.hidden = !trigger.hidden
	}
}